Zend/Optimizer/zend_optimizer.c

```diff
@@ -891,16 +891,20 @@ zend_function *zend_optimizer_get_called_func(
 					&op_array->scope->function_table, method_name);
 				if (fbc) {
 					bool is_private = (fbc->common.fn_flags & ZEND_ACC_PRIVATE) != 0;
-					bool is_final = (fbc->common.fn_flags & ZEND_ACC_FINAL) != 0;
-					bool same_scope = fbc->common.scope == op_array->scope;
 					if (is_private) {
 						/* Only use private method if in the same scope. We can't even use it
 						 * as a prototype, as it may be overridden with changed signature. */
+						bool same_scope = fbc->common.scope == op_array->scope;
 						return same_scope ? fbc : NULL;
 					}
-					/* If the method is non-final, it may be overridden,
-					 * but only with a compatible method signature. */
-					*is_prototype = !is_final;
+					/* Prototype methods are potentially overridden. fbc still contains useful type information.
+					 * Some optimizations may not be applied, like inlining or inferring the send-mode of superfluous args.
+					 * A method cannot be overridden if the class or method is final. */
+					bool can_be_overridden = (fbc->common.fn_flags & ZEND_ACC_FINAL) == 0 &&
+											 (fbc->common.scope->ce_flags & ZEND_ACC_FINAL) == 0;
+					if (can_be_overridden) {
```

Comment:
It doesn't make sense to introduce ``can_be_overridden`` variable. The condition may be used directly in the following ``if`` statement.

Reply:
I'll change that thanks.
About anonymous classes, see https://github.com/php/php-src/pull/10467#discussion_r1110427157

Reply:
> I'll change that thanks. About anonymous classes, see [#10467 (comment)](https://github.com/php/php-src/pull/10467#discussion_r1110427157)

Ugh, this is like overriding constants values through reflection API in Java :)
Anyway, your patch is right now.